driver/src/orderbook/mod.rs

```diff
@@ -27,26 +29,46 @@ pub trait StableXOrderBookReading {
     fn get_auction_data(&self, index: U256) -> Result<(AccountState, Vec<Order>)>;
 }
 
-/// A simple implementation of the order book trait that forwards calls
-/// directly to the smart contract.
-pub struct StableXOrderBookReader<'a> {
+/// Implements the StableXOrderBookReading trait by using the underlying
+/// contract in a paginated way.
+/// This avoid hitting gas limits when the total amount of orders is large.
+pub struct PaginatedStableXOrderBookReader<'a> {
     contract: &'a dyn StableXContract,
+    page_size: u64,
+    web3: &'a Web3,
 }
 
-impl<'a> StableXOrderBookReader<'a> {
-    pub fn new(contract: &'a dyn StableXContract) -> Self {
-        Self { contract }
+impl<'a> PaginatedStableXOrderBookReader<'a> {
+    pub fn new(contract: &'a dyn StableXContract, page_size: u64, web3: &'a Web3) -> Self {
+        Self {
+            contract,
+            page_size,
+            web3,
+        }
     }
 }
 
-impl<'a> StableXOrderBookReading for StableXOrderBookReader<'a> {
+impl<'a> StableXOrderBookReading for PaginatedStableXOrderBookReader<'a> {
     fn get_auction_index(&self) -> Result<U256> {
         self.contract
             .get_current_auction_index()
             .map(|batch_collecting_orders| batch_collecting_orders - 1)
     }
 
     fn get_auction_data(&self, index: U256) -> Result<(AccountState, Vec<Order>)> {
-        self.contract.get_auction_data(index)
+        let block = self.web3.eth().block_number().wait()?.as_u64();
+        let mut reader = BatchedAuctionDataReader::new(index);
+        loop {
+            let number_of_added_orders =
+                reader.apply_batch(&self.contract.get_auction_data_batched(
```

Comment:
Is this `apply_batch` doing something like `append_page`?

Reply:
I think this helps with the confusion of "batches" vs "pages" since the word "batch" has a very specific meaning in the exchange protocol.